### PR TITLE
Blockchain tests refactoring

### PR DESCRIPTION
**Pull request by winsvega**
based on blockchain test PR https://github.com/ethereum/cpp-ethereum/pull/2850
this one introduce new approach on blockchain tests
and separetes testing and mining logic

### CODE REVIEW

**Comment by CJentzsch** [comment]
is this PR complete (test filling works to 100%) or work in progress?

File: test/BlockChainHelper.h [hunk 1]
```diff
@@ -0,0 +1,128 @@
...
+public:
+	TestBlock() {};
+	TestBlock(TestBlock const& _original);	
+	TestBlock(std::string const& _blockRlp);
+	TestBlock& operator = (TestBlock const& _original);
+	TestBlock(mObject const& _blockObj, mObject const& _stateObj, RecalcBlockHeader _verify);
+	void addTransaction(TestTransaction const& _tr);
+	void addUncle(TestBlock const& _uncle);
+	void setUncles(vector<TestBlock> const& _uncles);
+	void mine(TestBlockChain const& bc);
```
Comment (CJentzsch) [inline test/BlockChainHelper.h]: _bc

File: test/BlockChainHelper.h [hunk 2]
```diff
@@ -0,0 +1,128 @@
...
+
+class TestBlockChain
+{
+private:
+	typedef FullBlockChain<Ethash> FullBlockChainEthash;
+public:
+	TestBlockChain(TestBlock const& _genesisBlock);	
+	void reset(TestBlock const& _genesisBlock);
+	void addBlock(TestBlock const& _block);
+	vector<TestBlock> syncUncles(vector<TestBlock> const& uncles);
```
Comment (CJentzsch) [inline test/BlockChainHelper.h]: _uncles

**Comment by CJentzsch** [comment]
It does not pass current tests:

Test Case "bcUncleTest": 
25%...
50%...
75%...
/home/christoph/software/Ethereum/cpp-ethereum/test/TestHelper.cpp(301): error in "bcUncleTest": Check State: 8888f1f195afa192cfee860698584c030f4c9db1: incorrect balance 15156250000000063000, expected 10000000000000042000
/home/christoph/software/Ethereum/cpp-ethereum/test/TestHelper.cpp(301): error in "bcUncleTest": Check State: 095e7baea6a6c7c4c2dfeb977efac326af552d87: incorrect balance 30, expected 20

File: test/BlockChainHelper.cpp [hunk 3]
```diff
@@ -0,0 +1,396 @@
...
+	m_testTransactions.clear();
+	for (auto const& tr: root[1])
+	{
+		Transaction tx(tr.data(), CheckTransaction::Everything);
+		TestTransaction testTx(tx);
+		m_transactionQueue.import(tx.rlp());
+		m_testTransactions.push_back(testTx);
+	}
+
+	for	(auto const& uRLP: root[2])
```
Comment (CJentzsch) [inline test/BlockChainHelper.cpp]: to much space

File: test/BlockChainHelper.cpp [hunk 4]
```diff
@@ -0,0 +1,396 @@
...
+		uBl.populateFromHeader(uRLP);
+		TestBlock uncle;
+		//uncle goes without transactions and uncles but
+		//it's hash could contain hashsum of transactions/uncles
+		//thus it won't need verification
+		uncle.setBlockHeader(uBl, RecalcBlockHeader::SkipVerify);
+		m_uncles.push_back(uncle);
+	}
+
+	//copyStateFrom(_original.getState());
```
Comment (CJentzsch) [inline test/BlockChainHelper.cpp]: remove comment?

File: test/BlockChainHelper.cpp [hunk 5]
```diff
@@ -0,0 +1,396 @@
...
+	}
+	catch (exception const& _e)
+	{
+		cnote << _e.what();
+	}
+}
+
+void TestBlock::addUncle(TestBlock const& _uncle)
+{
+	m_uncles.push_back(_uncle);
```
Comment (CJentzsch) [inline test/BlockChainHelper.cpp]: there a couple of functions only containing one line of code,why not use `m_uncles.push_back(_uncle);` directly instead of `addUncle(_uncle);` ?

**Comment by CJentzsch** [comment]
GasPricer tests are failing:

```
./testeth -t GasPricer
Running 12 test cases...
unknown location(0): fatal error in "basicGasPricer_bcValidBlockTest": std::runtime_error: value type is 6 not 0
/home/christoph/software/Ethereum/cpp-ethereum/test/libethereum/gaspricer.cpp(48): last checkpoint

*** 1 failure detected in test suite "Master Test Suite"
```

File: test/BlockChainHelper.cpp [hunk 6]
```diff
@@ -0,0 +1,396 @@
...
+}
+
+void TestBlock::mine(TestBlockChain const& bc)
+{
+	TestBlock const& genesisBlock = bc.getTestGenesis();
+	OverlayDB const& genesisDB = genesisBlock.getState().db();
+
+	FullBlockChain<Ethash> const& blockchain = bc.getInterface();
+
+	Block block = (blockchain.genesisBlock(genesisDB));
```
Comment (CJentzsch) [inline test/BlockChainHelper.cpp]: you don't the last pair of braces

File: test/BlockChainHelper.cpp [hunk 7]
```diff
@@ -0,0 +1,396 @@
...
+	RLPStream blHeaderStream;
+	m_blockHeader.streamRLP(blHeaderStream, WithProof);
+
+	RLPStream ret(3);
+	ret.appendRaw(blHeaderStream.out()); //block header
+	ret.appendRaw(txStream.out());		 //transactions
+	ret.appendRaw(uncleStream.out());	 //uncles
+
+	if (_recalculate == RecalcBlockHeader::Verify || _recalculate == RecalcBlockHeader::UpdateAndVerify)
+		m_blockHeader.verifyInternals(&ret.out());
```
Comment (CJentzsch) [inline test/BlockChainHelper.cpp]: the reason to verifyInternals is to check wether it throws or not. Why are you checking it here, or why don't you have a tr/catch around it?

File: test/BlockChainHelper.cpp [hunk 8]
```diff
@@ -0,0 +1,396 @@
...
+	m_bytes = ret.out();
+}
+
+void TestBlock::copyStateFrom(State const& _state)
+{
+	//WEIRD WAY TO COPY STATE AS COPY CONSTRUCTOR FOR STATE NOT IMPLEMENTED CORRECTLY (they would share the same DB)
```
Comment (CJentzsch) [inline test/BlockChainHelper.cpp]: than it would be better to fix the state copy constructor

File: test/BlockChainHelper.h [hunk 9]
```diff
@@ -0,0 +1,128 @@
...
+	You should have received a copy of the GNU General Public License
+	along with cpp-ethereum.  If not, see <http://www.gnu.org/licenses/>.
+*/
+/** @file BlockChainHelper.h
+ * @author Dimitry Khokhlov <dimitry@ethdev.com>
+ * @date 2015
+ */
+
+#pragma once
+#ifndef BLOCKCHAIN_HELPER_HEADER_H
```
Comment (CJentzsch) [inline test/BlockChainHelper.h]: thats what the pragma is for, why do you have it here?

File: test/BlockChainHelper.h [hunk 10]
```diff
@@ -0,0 +1,128 @@
...
+
+class TestTransaction
+{
+public:
+	TestTransaction(json_spirit::mObject const& _o);
+	TestTransaction(Transaction const& _tr) : m_transaction(_tr) {}
+	Transaction const& getTransaction() const { return m_transaction; }
+private:
+	json_spirit::mObject m_jsonTransaction;
+	Transaction  m_transaction;
```
Comment (CJentzsch) [inline test/BlockChainHelper.h]: looks like an additional space before m_transaction

File: test/BlockChainHelper.h [hunk 11]
```diff
@@ -0,0 +1,128 @@
...
+	Address const& getBeneficiary() const { return m_blockHeader.beneficiary(); }
+
+private:
+	BlockHeader constructBlock(mObject const& _o, h256 const& _stateRoot);
+	bytes createBlockRLPFromFields(mObject const& _tObj, h256 const& _stateRoot = h256{});
+	void recalcBlockHeaderBytes(RecalcBlockHeader _recalculate);
+	void copyStateFrom(State const& _state);
+	void populateFrom(TestBlock const& _original);
+
+private:
```
Comment (CJentzsch) [inline test/BlockChainHelper.h]: already declared private a couple of lines above

File: test/TestHelper.cpp [hunk 12]
```diff
@@ -890,4 +901,42 @@ void Listener::notifyTestFinished()
 		g_listener->testFinished();
 }
 
+size_t TestOutputHelper::m_currTest = 0;
```
Comment (CJentzsch) [inline test/TestHelper.cpp]: global (this file only) variables? Shouldn't they belong to a class?

File: test/TestHelper.cpp [hunk 13]
```diff
@@ -890,4 +901,42 @@ void Listener::notifyTestFinished()
...
+size_t TestOutputHelper::m_maxTests = 0;
+string TestOutputHelper::m_currentTestName = "n/a";
+
+using namespace boost;
+void TestOutputHelper::initTest(json_spirit::mValue& _v)
+{
+	std::string testCaseName = boost::unit_test::framework::current_test_case().p_name;
+	std::cerr << "Test Case \"" + testCaseName + "\": " << std::endl;
```
Comment (CJentzsch) [inline test/TestHelper.cpp]: why `std::err` , it doesn't sound like a error message

File: test/TestHelper.cpp [hunk 14]
```diff
@@ -890,4 +901,42 @@ void Listener::notifyTestFinished()
...
+bool TestOutputHelper::passTest(json_spirit::mObject& _o, std::string& _testName)
+{
+	m_currTest++;	
+	int m_testsPerProgs = std::max(1, (int)(m_maxTests / 4));
+	if (m_currTest % m_testsPerProgs == 0 || m_currTest ==  m_maxTests)
+	{
+		int percent = int(m_currTest*100/m_maxTests);
+		std::cerr << percent << "%";
+		if (percent != 100)
+			std::cerr << "...";
```
Comment (CJentzsch) [inline test/TestHelper.cpp]: same

File: test/libethereum/BlockchainTestsFiller/bcGasPricerTestFiller.json [hunk 15]
```diff
@@ -30,7 +33,7 @@
                 "storage": {}
             },
             "aaaf5374fce5edbc8e2a8697c15331677e6ebf0b" : {
-                "balance" : "10000000000",
```
Comment (CJentzsch) [inline test/libethereum/BlockchainTestsFiller/bcGasPricerTestFiller.json]: When you change something different from the `expect` section, please explain why.

**Comment by CJentzsch** [comment]
I saw that a lot of BlockChainTestfillers have changed. Except of the `expect` section, are those changes neccessary? Would filling still work with the current fillers?

File: test/libethereum/blockchain.cpp [hunk 16]
```diff
@@ -14,20 +14,22 @@
 	You should have received a copy of the GNU General Public License
 	along with cpp-ethereum.  If not, see <http://www.gnu.org/licenses/>.
 */
-/** @file block.cpp
- * @author Christoph Jentzsch <cj@ethdev.com>
+/** @file blockchain.cpp
+ * @author Dimitry Khokhlov <dimitry@ethdev.com>
```
Comment (CJentzsch) [inline test/libethereum/blockchain.cpp]: just leave us two as authors here :-)

File: test/libethereum/blockchain.cpp [hunk 17]
```diff
@@ -14,20 +14,22 @@
...
 #include <boost/filesystem.hpp>
+
+#include <test/TestHelper.h>
+#include <test/BlockChainHelper.h>
+#include <test/JsonSpiritHeaders.h>
+#include "test/fuzzTesting/fuzzHelper.h"
+
 #include <libdevcore/FileSystem.h>
-#include <libdevcore/TransientDirectory.h>
 #include <libethcore/Params.h>
```
Comment (CJentzsch) [inline test/libethereum/blockchain.cpp]: include order:  (C, STL, Boost, other libs, base ethereum libs, current ethereum lib)

File: test/libethereum/blockchain.cpp [hunk 18]
```diff
@@ -36,86 +38,45 @@ using namespace dev::eth;
...
-mObject writeBlockHeaderToJson(mObject& _o, BlockHeader const& _bi);
-void overwriteBlockHeader(BlockHeader& _current_BlockHeader, mObject& _blObj, const BlockHeader& _parent);
-void updatePoW(BlockHeader& _bi);
-mArray importUncles(mObject const& _blObj, vector<BlockHeader>& _vBiUncles, vector<BlockHeader> const& _vBiBlocks, std::vector<blockSet> _blockSet);
+//Functions that working with test json
+void compareBlocks(TestBlock const& _a, TestBlock const& _b);
+mArray writeTransactionsToJson(TransactionQueue const& _txsQueue);
+mObject writeBlockHeaderToJson(Ethash::BlockHeader const& _bi);
+void overwriteBlockHeaderForTest(mObject const& _blObj, TestBlock& _block, std::vector<TestBlock> const& importedBlocks);
+void overwriteUncleHeaderForTest(mObject& _uncleHeaderObj, TestBlock& _uncle, std::vector<TestBlock> const& _uncles, std::vector<TestBlock> const& _importedBlocks);
```
Comment (CJentzsch) [inline test/libethereum/blockchain.cpp]: since you are `using namespace std` you don't need that in this file (at several places)

File: test/libethereum/blockchain.cpp [hunk 19]
```diff
@@ -124,632 +85,425 @@ void doBlockchainTests(json_spirit::mValue& _v, bool _fillin)
...
-				FullBlockChain<Ethash> bc(rlpGenesisBlock.out(), AccountMap(), tdBc.path(), WithExisting::Kill);
-
-				//OverlayDB database (State::openDB(td_stateDB.path(), h256{}, WithExisting::Kill));
-				State state(OverlayDB(State::openDB(tdStateDB.path(), h256{}, WithExisting::Kill)), BaseState::Empty); //= importer.m_statePre;
-				ImportTest::importState(o["pre"].get_obj(), state);
-				state.commit();
-
-				//import previous blocks
-				for (size_t i = 1; i < importBlockNumber; i++) //0 block is genesis
+				if ((string)boost::unit_test::framework::current_test_case().p_name != "bcWalletTest")
```
Comment (CJentzsch) [inline test/libethereum/blockchain.cpp]: Still need to fix that problem. Because this just happens to be the only big BlockChainTest, but may be there will be reasons for other rather large BlockChainTests

File: test/libethereum/blockchain.cpp [hunk 20]
```diff
@@ -124,632 +85,425 @@ void doBlockchainTests(json_spirit::mValue& _v, bool _fillin)
...
+					cnote << testname + "block import throw an exception: " << _e.what();
+					wasException = true;
 				}
-				// if exception is thrown, RLP is invalid and no blockHeader, Transaction list, or Uncle list should be given
 				catch (...)
 				{
+					wasException = true;
+				}
+
+				if (wasException)
```
Comment (CJentzsch) [inline test/libethereum/blockchain.cpp]: instead of using the `wasException` bool, its nicer to write a little helper function and call that in the `catch` block

File: test/libethereum/blockchain.cpp [hunk 21]
```diff
@@ -124,632 +85,425 @@ void doBlockchainTests(json_spirit::mValue& _v, bool _fillin)
...
-				blockHeaderFromFields.populateFromHeader(c_blockHeaderRLP, IgnoreSeal);
+				//Check Provided Header against block in RLP
+				mObject emptyState;
+				TestBlock blockFromFields(blObj["blockHeader"].get_obj(), emptyState, RecalcBlockHeader::SkipVerify); //not verify as we havent imported transactions and uncles yet
 
-				BlockHeader blockFromRlp(trueBc.header());
-
-				if (importedAndBest)
+				//ImportTransactions
+				BOOST_REQUIRE(blObj.count("transactions"));
```
Comment (CJentzsch) [inline test/libethereum/blockchain.cpp]: additional spaces in the beginning

File: test/libethereum/blockchain.cpp [hunk 22]
```diff
@@ -124,632 +85,425 @@ void doBlockchainTests(json_spirit::mValue& _v, bool _fillin)
...
-			BOOST_CHECK_MESSAGE(toString(trueBc.info().hash()) == o["lastblockhash"].get_str(),
-					testname + "Boost check: " + i.first + " lastblockhash does not match " + toString(trueBc.info().hash()) + " expected: " + o["lastblockhash"].get_str());
+				try
+				{
+					blockFromFields.setBlockHeader(blockFromFields.getBlockHeader(), RecalcBlockHeader::Verify); //recalculateBytes
+					blockchain.addBlock(blockFromFields);
+				}
+				catch (Exception const& _e)
+				{
+					cerr << "Error importing block from filds to blockchain: " << diagnostic_information(_e);
```
Comment (CJentzsch) [inline test/libethereum/blockchain.cpp]: typo

File: test/libethereum/blockchain.cpp [hunk 23]
```diff
@@ -124,632 +85,425 @@ void doBlockchainTests(json_spirit::mValue& _v, bool _fillin)
...
+	_block.setBlockHeader(tmp, findNewValidNonce);
 }
 
-bytes createBlockRLPFromFields(mObject& _tObj, h256 const& _stateRoot)
+void overwriteUncleHeaderForTest(mObject& uncleHeaderObj, TestBlock& uncle, std::vector<TestBlock> const& uncles, std::vector<TestBlock> const& importedBlocks)
 {
-	RLPStream rlpStream;
-	rlpStream.appendList(_tObj.count("hash") > 0 ? (_tObj.size() - 1) : _tObj.size());
-
-	if (_tObj.count("parentHash"))
```
Comment (CJentzsch) [inline test/libethereum/blockchain.cpp]: why did you remove the ability to overwrite singe or all fields individually?

File: test/libethereum/blockchain.cpp [hunk 24]
```diff
@@ -124,632 +85,425 @@ void doBlockchainTests(json_spirit::mValue& _v, bool _fillin)
...
+	}
 
-void overwriteBlockHeader(BlockHeader& _header, mObject& _blObj, BlockHeader const& _parent)
-{
-	auto ho = _blObj["blockHeader"].get_obj();
-	if (ho.size() != 14)
+	if (uncleHeaderObj.count("sameAsPreviousBlockUncle"))
 	{
-		BlockHeader tmp = constructHeader(
-			ho.count("parentHash") ? h256(ho["parentHash"].get_str()) : _header.parentHash(),
```
Comment (CJentzsch) [inline test/libethereum/blockchain.cpp]: Same question as above. I want the old fillers to still work.

File: test/libethereum/blockchain.cpp [hunk 25]
```diff
@@ -806,16 +608,9 @@ BOOST_AUTO_TEST_CASE(bcTotalDifficultyTest)
 	dev::test::executeTests("bcTotalDifficultyTest", "/BlockchainTests",dev::test::getFolder(__FILE__) + "/BlockchainTestsFiller", dev::test::doBlockchainTests);
 }
 
-BOOST_AUTO_TEST_CASE(bcForkUncleTest)
```
Comment (CJentzsch) [inline test/libethereum/blockchain.cpp]: why do you remove this test?

File: test/libethereum/blockchain.cpp [hunk 26]
```diff
@@ -757,48 +511,96 @@ mArray writeTransactionsToJson(Transactions const& _txs)
...
 }
 
-} }// Namespace Close
+}}//namespaces
 
 BOOST_AUTO_TEST_SUITE(BlockChainTests)
 
 BOOST_AUTO_TEST_CASE(bcForkBlockTest)
 {
-	if (!test::Options::get().fillTests)
```
Comment (CJentzsch) [inline test/libethereum/blockchain.cpp]: I did this to prevent errors when filling BlockChainTests which don't have a filler, how do you handle this situation now?

File: test/BlockChainHelper.cpp (hunk 5 above)
Comment (winsvega) [inline test/BlockChainHelper.cpp]: m_uncles is a private field

File: test/BlockChainHelper.cpp (hunk 8 above)
Comment (winsvega) [inline test/BlockChainHelper.cpp]: we could try it later and discuss with Gavin

File: test/TestHelper.cpp (hunk 12 above)
Comment (winsvega) [inline test/TestHelper.cpp]: this do belong to a class. TestOutputHelper. 
a static class for handling output of the tests.

File: test/TestHelper.cpp (hunk 13 above)
Comment (winsvega) [inline test/TestHelper.cpp]: with verbosity 1 only error messages are thrown. But thus it is kind of unclear wethere tests are doing ok, or they stack at some point. Tats why I added output to show that tests are working.

File: test/libethereum/blockchain.cpp (hunk 19 above)
Comment (winsvega) [inline test/libethereum/blockchain.cpp]: that would be a trouble. cause I keep history of all imported blocks with it states
thus blockchain could be rebuld from any block for the test
changes might require to reconsider this logic while quicker solution is add some test option to disable it.  
but if there will be tests with blockchain overwrite at block like #100  this would be a problem

File: test/libethereum/blockchain.cpp (hunk 23 above)
Comment (winsvega) [inline test/libethereum/blockchain.cpp]: it moved to overwriteBlockHeaderForTest.

File: test/TestHelper.cpp (hunk 13 above)
Comment (CJentzsch) [inline test/TestHelper.cpp]: Yes to output, but no for using std::err for that. You can make or use a different logging chanel for that which does work with verbosity 1.

File: test/libethereum/blockchain.cpp (hunk 19 above)
Comment (CJentzsch) [inline test/libethereum/blockchain.cpp]: It does work with the previous blockchain test logic, although it did take a while and was inefficient. But it did not crash, the same would be sufficient for now.

File: test/libethereum/blockchain.cpp (hunk 24 above)
Comment (winsvega) [inline test/libethereum/blockchain.cpp]: old fillers are working but I did some changes to them to ensure that they working correctly,

File: test/libethereum/blockchain.cpp (hunk 19 above)
Comment (winsvega) [inline test/libethereum/blockchain.cpp]: before it was one State object for all blocks. 
Now I keep idicidual State for each block. it crush if too many of them are open

File: test/TestHelper.cpp (hunk 13 above)
Comment (winsvega) [inline test/TestHelper.cpp]: I will use cout for that

File: test/libethereum/blockchain.cpp (hunk 21 above)
Comment (winsvega) [inline test/libethereum/blockchain.cpp]: ?

File: test/libethereum/blockchain.cpp (hunk 26 above)
Comment (winsvega) [inline test/libethereum/blockchain.cpp]: restored this back

File: test/libethereum/blockchain.cpp (hunk 21 above)
Comment (CJentzsch) [inline test/libethereum/blockchain.cpp]: On github it looks like you have used spaces instead of tabs here.

File: test/BlockChainHelper.cpp (hunk 8 above)
Comment (CJentzsch) [inline test/BlockChainHelper.cpp]: confirmed, I had locally a different BlockchainTests/bcValidBlockTest.json file which caused that

File: test/BlockChainHelper.h [hunk 27]
```diff
@@ -0,0 +1,126 @@
...
+	BlockQueue m_uncleQueue;
+	bytes m_bytes;
+	std::unique_ptr<TransientDirectory> m_tempDirState;
+	vector<TestTransaction> m_testTransactions;
+};
+
+class TestBlockChain
+{
+private:
+	typedef FullBlockChain<Ethash> FullBlockChainEthash;
```
Comment (CJentzsch) [inline test/BlockChainHelper.h]: Tthis can be put under the other privat section. There should be only on per class.

**Comment by CJentzsch** [comment]
code looks good, but filling the tests exactly the same way the old code did does not work 100%. when this works, it can be merged.
